File: src/main/java/com/itdhq/maven/aconfgen/plugin/GenerateConstantsMojo.java
package com.itdhq.maven.aconfgen.plugin;

import freemarker.template.Configuration;
import freemarker.template.Template;
import freemarker.template.TemplateException;
import org.alfresco.repo.dictionary.*;
import org.alfresco.service.cmr.dictionary.*;
import org.alfresco.service.namespace.NamespaceService;
import org.alfresco.service.namespace.QName;
import org.apache.maven.plugin.MojoExecutionException;
import org.apache.maven.plugin.MojoFailureException;
import org.apache.maven.plugins.annotations.LifecyclePhase;
import org.apache.maven.plugins.annotations.Mojo;
import org.apache.maven.plugins.annotations.Parameter;
import org.springframework.context.ApplicationContext;
import org.springframework.context.support.ClassPathXmlApplicationContext;

import java.io.*;
import java.util.*;

@Mojo(name = "gen-const", defaultPhase = LifecyclePhase.GENERATE_SOURCES, threadSafe = true)
public class GenerateConstantsMojo extends AbstractAconfgenMojo {

    private final String PROPERTY = "PROP";

    private final String ASSOCIATION = "ASSOC";

    private final String CONSTRAINT = "CONSTRAINT";

    private ApplicationContext context;

    private String packageName;

    private String className;

    private DictionaryDAO dictionaryDAO;

    private NamespaceDAO namespaceDAO;

    private Map<String, String> uriNamesCache = new HashMap<String, String>();

    private Map<QName, Collection<String>> constraintsCache = new HashMap<QName, Collection<String>>();

    private Set<NameParamsConstant> qNameConstants;

    private Set<NameValueConstant> stringConstants;

    private Set<QName> dependencyConstraints = new HashSet<QName>();

    @Parameter(defaultValue = "${project.build.directory}/generated-sources/aconfgen/")
    private File outputDirectory;

    @Parameter
    private Constants[] constants;

    private Set<String> excludedModels = new HashSet<String>(Arrays.asList(
            new String[]{
                    NamespaceService.WORKFLOW_MODEL_1_0_URI,
                    NamespaceService.WEBDAV_MODEL_1_0_URI,
                    NamespaceService.SYSTEM_MODEL_1_0_URI,
                    NamespaceService.SECURITY_MODEL_1_0_URI,
                    NamespaceService.LINKS_MODEL_1_0_URI,
                    NamespaceService.EXIF_MODEL_1_0_URI,
                    NamespaceService.FORUMS_MODEL_1_0_URI,
                    NamespaceService.EMAILSERVER_MODEL_URI,
                    NamespaceService.DICTIONARY_MODEL_1_0_URI,
                    NamespaceService.DATALIST_MODEL_1_0_URI,
                    NamespaceService.CONTENT_MODEL_1_0_URI,
                    NamespaceService.BPM_MODEL_1_0_URI,
                    NamespaceService.AUDIO_MODEL_1_0_URI,
                    NamespaceService.APP_MODEL_1_0_URI
            }
    ));

    public void execute() throws MojoExecutionException, MojoFailureException {

        context = new ClassPathXmlApplicationContext("data-model-context.xml");
        dictionaryDAO = (DictionaryDAO) context.getBean("dictionaryDAO");
        namespaceDAO = (NamespaceDAO) context.getBean("namespaceDAO");

        processCommonModels();

        for (Constants constant : constants) {

            setNames(constant);

            qNameConstants = new HashSet<NameParamsConstant>();
            stringConstants = new HashSet<NameValueConstant>();

            for (String path : constant.getModels()) {

                CompiledModel compiledModel;
                try {
                   compiledModel = getCompiledModel(path);
                }
                catch (FileNotFoundException e) {
                    getLog().error("Model " + path + " could not be read");
                    break;
                }

                QName modelName = compiledModel.getModelDefinition().getName();
                dictionaryDAO.putModel(compiledModel.getM2Model());

                for (NamespaceDefinition namespaceDefinition : compiledModel.getModelDefinition().getNamespaces()) {
                    stringConstants.add(createUri(modelName, namespaceDefinition.getUri()));
                    stringConstants.add(createPrefix(modelName));
                }

                for (ConstraintDefinition constraintDefinition : compiledModel.getConstraints()) {
                    Constraint constraint = constraintDefinition.getConstraint();
                    QName constraintQName = QName.createQName(constraint.getShortName(), namespaceDAO);
                    if (!constraintsCache.containsKey(constraintQName) || dependencyConstraints.contains(constraintQName)) {
                        qNameConstants.add(createConstantQName(constraintQName, CONSTRAINT));
                        if (constraint.getType().equals("LIST")) {
                            Collection<String> values = (Collection<String>) constraint.getParameters()
                                    .get("allowedValues");
                            constraintsCache.put(constraintQName, values);
                        }
                    }

                }


                for (TypeDefinition typeDefinition : compiledModel.getTypes()) {
                    processProperties(typeDefinition.getProperties());

                    for (QName qName : typeDefinition.getAssociations().keySet()) {
                        if (isAllowed(qName)) {
                            qNameConstants.add(createConstantQName(qName, ASSOCIATION));
                        }
                    }
                    for (AspectDefinition aspectDefinition : typeDefinition.getDefaultAspects()) {
                        if (isAllowed(aspectDefinition.getName())) {
                            processProperties(aspectDefinition.getProperties());

                            for (QName qName : aspectDefinition.getAssociations().keySet()) {
                                if (isAllowed(qName)) {
                                    qNameConstants.add(createConstantQName(qName, ASSOCIATION));
                                }
                            }
                        }
                    }
                }
            }
            try {
                generateConstants();
            }
            catch (IOException e) {
                e.printStackTrace();
            }
            catch (TemplateException e) {
                getLog().error("Cannot write file to " + outputDirectory.getAbsolutePath());
            }
        }
    }

    private void processProperties(Map<QName, PropertyDefinition> properties) {
        for (QName qName : properties.keySet()) {
            if (isAllowed(qName)) {
                qNameConstants.add(createConstantQName(qName, PROPERTY));
                PropertyDefinition property = properties.get(qName);
                for (ConstraintDefinition constraintDefinition : property.getConstraints()) {
                    Constraint constraint = constraintDefinition.getConstraint();
                    QName constraintQName = QName.createQName(constraint.getShortName(), namespaceDAO);
                    if (constraintsCache.containsKey(constraintQName)) {
                        for (String value : constraintsCache.get(constraintQName)) {
                            stringConstants.add(createPropertyValues(qName, value));
                        }
                    }
                }
            }
        }
    }

    private CompiledModel getCompiledModel(String path) throws FileNotFoundException {
        InputStream inputStream = new FileInputStream(path);
        M2Model model = M2Model.createModel(inputStream);
        return model.compile(dictionaryDAO, namespaceDAO, true);
    }

    private boolean isAllowed(QName qName) {
        return !excludedModels.contains(qName.getNamespaceURI());
    }

    private void setNames(Constants constant) {
        String path = constant.getPath();
        if (path != null) {
            if (path.endsWith(".java")) {
                path.replace(".java", "");
            }
            int splitIndex = constant.getPath().lastIndexOf('.');
            packageName = path.substring(0, splitIndex);
            className = path.substring(splitIndex + 1, path.length());
        }
    }

    private NameValueConstant createUri(QName qName, String uri) {
        String name = getPrefix(qName.getPrefixString()).toUpperCase() + "_URI";
        uriNamesCache.put(uri, name);
        return new NameValueConstant(name, uri);
    }

    private NameValueConstant createPrefix(QName qName) {
        String name = getPrefix(qName.getPrefixString());
        return new NameValueConstant(name.toUpperCase() + "_PREFIX", name);
    }

    private NameValueConstant createPropertyValues(QName qName, String value) {
        String name = createName(qName, PROPERTY);
        String postfix = value.replace(' ', '_').replace('-', '_');
        if (value.equals("")) {
            name += "_EMPTY";
        }
        else {
            name += "_" + postfix.toUpperCase();
        }
        return new NameValueConstant(name, value);
    }

    private String getPrefix(String line) {
        if (line.contains(":")) {
            return line.substring(0, line.indexOf(':'));
        }
        return line;
    }

    private NameParamsConstant createConstantQName(QName qName, String prefix) {
        String name = createName(qName, prefix);
        String uri;
        if (uriNamesCache.containsKey(qName.getNamespaceURI())) {
            uri = uriNamesCache.get(qName.getNamespaceURI());
        }
        else
            uri = "\"" + qName.getNamespaceURI() + "\"";
        String localName = qName.getLocalName();
        return new NameParamsConstant(name, uri, localName);
    }

    private String createName(QName qName, String prefix) {
        String prefixString = qName.getPrefixString();
        String shortName = qName.getLocalName();
        String[] words = shortName.split("(?=\\p{Lu})");
        StringBuilder name = new StringBuilder(prefix + "_" + getPrefix(prefixString)
                .toUpperCase() + "_");
        for (String word : words) {
            name.append(word.toUpperCase()).append("_");
        }
        return name.substring(0, name.length() - 1);
    }

    private void generateConstants() throws IOException, TemplateException {
        Configuration configuration = new Configuration();
        configuration.setClassForTemplateLoading(this.getClass(), "/templates/");
        configuration.setDefaultEncoding("UTF-8");
        Map<String, Object> input = new HashMap<String, Object>();
        input.put("packageName", packageName);
        input.put("name", className);
        input.put("paramsConstants", qNameConstants);
        input.put("stringConstants", stringConstants);
        packageName = packageName.replace('.', '/');
        outputDirectory = new File(outputDirectory + "/" + packageName);
        outputDirectory.mkdirs();
        project.addCompileSourceRoot(outputDirectory.getAbsolutePath());
        Template template = configuration.getTemplate("constants-class.ftl");
        Writer fileWriter = new FileWriter(new File(outputDirectory + "/" + className + ".java"));
        template.process(input, fileWriter);
        fileWriter.close();
    }

    @Override
    protected void processCommonModels() {
        String[] models = getModels();
        for (String model : models) {
            CompiledModel compiledModel;
            try {
                compiledModel = getCompiledModel(model);
            }
            catch (FileNotFoundException e) {
                getLog().error("Model " + model + " could not be read");
                break;
            }
            dictionaryDAO.putModel(compiledModel.getM2Model());
            for (ConstraintDefinition constraintDefinition : compiledModel.getConstraints()) {
                Constraint constraint = constraintDefinition.getConstraint();
                if (constraint.getType().equals("LIST")) {
                    Collection<String> values = (Collection<String>) constraint.getParameters()
                            .get("allowedValues");
                    QName constraintQName = QName.createQName(constraint.getShortName(), namespaceDAO);
                    if (!dependencyConstraints.contains(constraintQName)) {
                        dependencyConstraints.add(constraintQName);
                    }
                    constraintsCache.put(constraintQName, values);
                }
            }
        }

    }
}

File: src/main/kotlin/com/itdhq/maven/aconfgen/plugin/GenerateConstantsMojo.kt
package com.itdhq.maven.aconfgen.plugin.kotlin

import com.itdhq.maven.aconfgen.plugin.freemarker.GetPropertyAllowedValues
import com.itdhq.maven.aconfgen.plugin.freemarker.GetQNamePrefix
import com.itdhq.maven.aconfgen.plugin.freemarker.SnakeUpperCase
import freemarker.template.Configuration
import org.alfresco.repo.dictionary.DictionaryDAO
import org.alfresco.repo.dictionary.M2Model
import org.alfresco.repo.dictionary.NamespaceDAO
import org.alfresco.service.namespace.QName
import org.apache.maven.plugin.AbstractMojo
import org.apache.maven.plugins.annotations.LifecyclePhase
import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import org.apache.maven.project.MavenProject
import org.apache.maven.shared.model.fileset.FileSet
import org.apache.maven.shared.model.fileset.util.FileSetManager
import org.springframework.context.support.ClassPathXmlApplicationContext
import java.io.File
import java.io.FileInputStream
import java.io.FileWriter
import java.util.*

class ConstantsItem {

    @Parameter
    lateinit var qualifiedName: String

    @Parameter
    lateinit var models: FileSet
}

@Mojo(name = "generate-constants", defaultPhase = LifecyclePhase.GENERATE_SOURCES, threadSafe = true)
class GenerateConstantsMojo : AbstractMojo() {

    @Parameter(defaultValue = "\${project.build.directory}/generated-sources/aconfgen/")
    private lateinit var outputDirectory: File

    @Parameter
    private lateinit var constants: List<ConstantsItem>

    @Parameter(defaultValue = "\${project}")
    private val project: MavenProject? = null


    private val configuration = Configuration()

    private lateinit var namespaceDAO: NamespaceDAO
    private lateinit var dictionaryDAO: DictionaryDAO

    override fun execute() {
        val context = ClassPathXmlApplicationContext("data-model-context.xml")
        dictionaryDAO = context.getBean("dictionaryDAO") as DictionaryDAO
        namespaceDAO = context.getBean("namespaceDAO") as NamespaceDAO

        loadFreeMarkerConfig()

        try {
            constants.forEach { generateConstants(it, dictionaryDAO) }
        } catch (e: Throwable) {
            log.error("Failed to generate constants: ${e.message}", e)
        }

        project!!.addCompileSourceRoot(outputDirectory.absolutePath)
    }

    private fun loadFreeMarkerConfig() {
        configuration.setClassForTemplateLoading(this.javaClass, "/templates/")
        configuration.defaultEncoding = "UTF-8"

        configuration.setSharedVariable("get_qname_ns_prefix", GetQNamePrefix(namespaceDAO))
        configuration.setSharedVariable("snake_upper_case", SnakeUpperCase())
        configuration.setSharedVariable("get_prop_allowed_values", GetPropertyAllowedValues())
    }

    private fun generateConstants(item: ConstantsItem, dictionaryDAO: DictionaryDAO) {
        val fileSetManager = FileSetManager()

        val queue = fileSetManager.getIncludedFiles(item.models)
                .map { FileInputStream(File(item.models.directory, it)) }
                .mapTo(LinkedList<M2Model>()) { M2Model.createModel(it) }

        val loadedModels = LinkedList<QName>()

        var counter = 0
        while (queue.isNotEmpty()) {
            if (counter == queue.size)
                throw Exception("Failed to load models. Please check that all dependencies specified.")

            val model = queue.pop()

            try {
                loadedModels.add(dictionaryDAO.putModel(model))
                counter = 0
            } catch (e: Exception) {
                queue.add(model)
                counter++
            }
        }

        val pos = item.qualifiedName.lastIndexOf('.')
        if (pos == -1)
            throw Exception("Qualified name ${item.qualifiedName} is invalid")

        val packageName = item.qualifiedName.substring(0, pos)
        val className = item.qualifiedName.substring(pos + 1)

        var dest = packageName.split('.').fold(outputDirectory, { acc, s -> File(acc, s) })
        dest.mkdirs()

        dest = File(dest, "${className}.java")

        val templateModel = hashMapOf(
                "packageName" to packageName,
                "className" to className,
                "namespaces" to loadedModels.flatMap { dictionaryDAO.getModel(it).namespaces },
                "types" to loadedModels.flatMap(dictionaryDAO::getTypes),
                "properties" to loadedModels.flatMap(dictionaryDAO::getProperties),
                "associations" to loadedModels.flatMap(dictionaryDAO::getAssociations)
        )

        configuration.getTemplate("constants-class.ftl").process(templateModel, FileWriter(dest))
    }

}